### PR TITLE
Add regex and GoofyAPI

### PR DESCRIPTION
**Pull request by TheBunnyMan123**

### CODE REVIEW

File: common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        overloads = {
+            @LuaMethodOverload(
+                argumentTypes = {String.class, String.class},
+                argumentNames = {"str", "pattern"}
+            )
+        },
+        value = "goofy.regex_match"
+    )
+    public String[] regexMatch(@LuaNotNil String str, @LuaNotNil String pattern) {
+        Pattern goofyPattern = Pattern.compile(pattern);
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java:37]: could some sort of caching be used here?

File: common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        Matcher matcher = goofyPattern.matcher(str);
+        
+        ArrayList<String> matches = new ArrayList<String>();
+
+        while (matcher.find()) {
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java:42]: a `gmatch`-like API would also be cool but I'll implement that tonight

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraAPIManagerMixin.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    @Shadow
+    @Final
+    public static Map<String, Function<FiguraLuaRuntime, Object>> API_GETTERS;
+
+    static {
+        WHITELISTED_CLASSES.add(GoofyAPI.class);
+
+        API_GETTERS.put("goofy", r -> new GoofyAPI(r));
+    }
+}
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraAPIManagerMixin.java:54]: no ending EOL; some programs will complain
```suggestion
}
```

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraDocsManagerMixin.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @Final
+    public static Map<String, Collection<Class<?>>> GLOBAL_CHILDREN;
+
+
+    static {
+        GLOBAL_CHILDREN.put("goofy", List.of(
+            GoofyAPI.class
+        ));
+    }
+}
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraDocsManagerMixin.java:25]: no ending EOL; some programs will complain
```suggestion
}
```

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraGlobalsDocsMixin.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+import com.thekillerbunny.goofyplugin.lua.GoofyAPI;
+import org.figuramc.figura.lua.docs.FiguraGlobalsDocs;
+import org.figuramc.figura.lua.docs.LuaFieldDoc;
+import org.spongepowered.asm.mixin.Mixin;
+
+@Mixin(value = FiguraGlobalsDocs.class, remap = false)
+public class FiguraGlobalsDocsMixin {
+    @LuaFieldDoc("globals.goofy")
+    public GoofyAPI goofy;
+}
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraGlobalsDocsMixin.java:12]: no ending EOL; some programs will complain
```suggestion
}
```

File: common/src/main/resources/assets/exampleplugin/lang/en_us.json [hunk 6]
```diff
@@ -2,5 +2,9 @@
   "goofyfiguraplugin.gui.goofyscreen.ok.tooltip": "Press this to return to the previous screen",
   "goofyfiguraplugin.gui.goofyscreen": "Example Screen",
   "goofyfiguraplugin.permissions.value.example_permission": "Example Permission",
-  "goofyfiguraplugin.permissions.value.example_permission.tooltip": "This is an example for a permission"
+  "goofyfiguraplugin.permissions.value.example_permission.tooltip": "This is an example for a permission",
+  "figura.docs.goofy": "The dumping ground of the figura goofballs",
+  "figura.docs.globals.goofy": "The global instance of GoofyAPI",
+  "figura.docs.goofy.regex_match": "Returns a list of matches based on the specified\npattern.",
+  "figura.docs.goofy.regex_sub": "Returns a string with all matches of the specified\npattern replaced."
 }
```
Comment (PoolloverNathan) [inline common/src/main/resources/assets/exampleplugin/lang/en_us.json:10]: no ending EOL; some programs will complain
```suggestion
}
```

**Review by PoolloverNathan — CHANGES_REQUESTED** [review]
missing some EOLs; should I add a check for this?

File: common/src/main/resources/assets/exampleplugin/lang/en_us.json (hunk 6 above)
Comment (PoolloverNathan) [inline common/src/main/resources/assets/exampleplugin/lang/en_us.json:10]:
```suggestion
}

```

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraDocsManagerMixin.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+public class FiguraDocsManagerMixin {
+    @Shadow
+    @Final
+    public static Map<String, Collection<Class<?>>> GLOBAL_CHILDREN;
+
+
+    static {
+        GLOBAL_CHILDREN.put("goofy", List.of(
+            GoofyAPI.class
+        ));
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraDocsManagerMixin.java:23]: is this line splitting necessary if the plan is to only add functions to `GoofyAPI`? (this is a question; you don't have to apply it)
```suggestion
        GLOBAL_CHILDREN.put("goofy", List.of(GoofyAPI.class));
```

File: common/src/main/resources/goofyplugin-common.mixins.json [hunk 8]
```diff
@@ -6,6 +6,10 @@
   "client": [
   ],
   "mixins": [
+    "FiguraAPIManagerMixin",
+    "FiguraDocsManagerMixin",
+    "FiguraGlobalsDocsMixin"
+
```
Comment (PoolloverNathan) [inline common/src/main/resources/goofyplugin-common.mixins.json:12]: unnecessary waste of line
```suggestion
```

**Review by PoolloverNathan — CHANGES_REQUESTED** [review]

**Review by PoolloverNathan — APPROVED** [review]

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraDocsManagerMixin.java (hunk 7 above)
Comment (TheBunnyMan123) [reply common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraDocsManagerMixin.java:23]: I like the look of it better

**Comment by PoolloverNathan** [comment]
ready for merging but I have a question (scroll up)

File: common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java (hunk 2 above)
Comment (TheBunnyMan123) [reply common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java:42]: I agree, gmatch would be great. that's on you after this is merged

**Comment by PoolloverNathan** [comment]
LGTM

File: common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java (hunk 1 above)
Comment (TheBunnyMan123) [reply common/src/main/java/com/thekillerbunny/goofyplugin/lua/GoofyAPI.java:37]: feel free to add it when you add gmatch-like function

**Pull request merged**